### ISSUE TITLE
Default resolver for interface fields

### ISSUE THREAD

**Issue by theorygeek** [rmosolgo/graphql-ruby#130]
In #108, we addressed the "mixin" type functionality that graphql-ruby was missing by allowing interfaces to provide default implementations for fields. This is an incredibly useful feature, and allows an interface to function similarly to how modules function in plain Ruby; I try to do this with all of my interfaces.

However, what that also means is that if you _don't_ provide an implementation for a field that comes from an interface, you'll get the default behavior, which is to do [`public_send`](https://github.com/rmosolgo/graphql-ruby/blob/765f1f6759003320183696178ed19152cd4641f4/lib/graphql/field.rb#L110) against the underlying object.

In my experience, this is almost never the correct behavior, and usually means that I simply forgot that the `ObjectType` I was creating needs to define that field. Many of the fields defined in my interfaces require more complicated resolver methods, and often the names of the methods are different anyway (`camelCase` vs `snake_case`).

So I think it might be good to require the developer to "opt-in" to receiving the default field implementations from interfaces, but that may be controversial, so I thought I'd just ask in an issue before doing a PR.

**Possible Solutions**
One solution would be to modify [`GraphQL::ObjectType`](https://github.com/rmosolgo/graphql-ruby/blob/master/lib/graphql/object_type.rb) so that it would only include fields from an interface if the interface was not using a default resolver. For example, the developer would have to define `resolve` proc, or specify a `property:` or `field:`, etc. If they did not do this, it would not be returned by [`#get_field`](https://github.com/rmosolgo/graphql-ruby/blob/4345c4808be0c1d0a74562da366d7a03e505c4ca/lib/graphql/object_type.rb#L49), or `#all_fields`, etc.

Another solution would be to change the default implementations for fields defined on interfaces to throw errors, although the drawback is that you wouldn't find out that you forgot to define a field until runtime.

A third solution would be to add some sort of flag to the field when it's defined in the interface, such as `inherit: true`, and only these fields would be copied to the object type.

What do you guys think?
/cc: @rmosolgo @xuorig @peterhorne

**Comment by rmosolgo** [comment]
I'm starting to sound like a broken record on this "accepts_definitions" API but here I go again! 

Here's an option you could use (or we could add to this gem):

```ruby 
# Define a callable definition
module AbstractField 
  def self.call(interface_type, field_name, return_type)
    field_name_s = field_name.to_s 
    abstract_field = GraphQL::Field.define do 
      name(field_name_s)
      type(return_type)
      resolve -> (obj, args, ctx) { raise(NotImplementedError, "Can't resolve #{field_name_s} on #{obj}; resolution not defined" }
    end 
    interface_type.fields[field_name_s] = abstract_field
  end
end

GraphQL::InterfaceType.accepts_definitions(abstract_field: AbstractField) 

MyInterface = GraphQL::InterfaceType.define do 
  # ... 
  abstract_field :something, types.String # will raise NotImplemented if accessed
end
```

This actually isn't quite right -- we want it to :boom: at load-time, not runtime when someone wants to use it... but that might be a start!

**Comment by rmosolgo** [comment]
One thing that needs an overhaul is `Schema`'s build-time validation. Right now the implementation is obtuse and it's missing some crucial features (validating interface implementation comes to mind :S). Maybe that can be redone in the same way `.define` was redone, so that it applies a collection of rules. Then, the library could _come_ with a bunch of default rules, but you could add your own too.

**Comment by rmosolgo** [comment]
FWIW I'd still like to support this, something like 

```ruby
field :name, abstract: true 
```

The choice to inherit by default might not have been the right one!

**Comment by rmosolgo** [comment]
I just left this open because I hadn't actually _done_ it yet! 😅

**Comment by theorygeek** [comment]
Oh no worries... I was honestly just cleaning up. I'll open it back up, that's a great suggestion!

**Comment by rmosolgo** [comment]
Removing inheritance-by-default with the `implements` method in #574

**Comment by rmosolgo** [comment]
This will be available via `implements` in 1.5.0 !

**Comment by theorygeek** [comment]
Nice, thanks! 🎉
